examples/core/rgcn/hetero_rgcn.py

```diff
@@ -0,0 +1,654 @@
+
+    (HIGHLIGHT)
+    A HeteroEmbedding instance holds separate embedding layers for each node
+    type, each with its own feature space of dimensionality
+    (node_num[i], embed_size), where 'node_num[i]' is the number of nodes of
+    type 'ntype' and 'embed_size' is the embedding dimension.
+
+    The "paper" node type is specifically excluded, possibly because these nodes
+    might already have predefined feature representations, and therefore, do not
+    require an additional embedding layer.
+
+    Parameters
+    ----------
+    graph : DGLGraph
+        The graph for which to create the heterogenous embedding layer.
+    embed_size : int
+        The size of the embedding vectors.
+
+    Returns
+    --------
+    HeteroEmbedding
+        A heterogenous embedding layer for all node types in the graph, except
+        for the "paper" node type.
+    """
+    node_num = {}
+    for ntype in graph.ntypes:
+        # Skip the "paper" node type.
+        if ntype == "paper":
+            continue
+        node_num[ntype] = graph.num_nodes(ntype)
+    embeds = HeteroEmbedding(node_num, embed_size)
+    return embeds
+
+
+class RelGraphConvLayer(nn.Module):
+    def __init__(
+        self,
+        in_size,
+        out_size,
+        ntypes,
+        relation_names,
+        activation=None,
+        dropout=0.0,
+    ):
+        super(RelGraphConvLayer, self).__init__()
+        self.in_size = in_size
+        self.out_size = out_size
+        self.ntypes = ntypes
+        self.relation_names = relation_names
+        self.activation = activation
+
+        ########################################################################
+        # (HIGHLIGHT) HeteroGraphConv is a graph convolution operator over
+        # heterogeneous graphs. A dictionary is passed where the key is the
+        # relation name and the value is the instance of GraphConv. norm="right"
+        # is to divide the aggregated messages by each node’s in-degrees, which
+        # is equivalent to averaging the received messages. weight=False and
+        # bias=False as we will use our own weight matrices defined later.
+        ########################################################################
+        self.conv = dglnn.HeteroGraphConv(
+            {
+                rel: dglnn.GraphConv(
+                    in_size, out_size, norm="right", weight=False, bias=False
+                )
+                for rel in relation_names
+            }
+        )
+
+        # Create a separate Linear layer for each relationship. Each
+        # relationship has its own weights which will be applied to the node
+        # features before performing convolution.
+        self.weight = nn.ModuleDict(
+            {
+                rel_name: nn.Linear(in_size, out_size, bias=False)
+                for rel_name in self.relation_names
+            }
+        )
+
+        # Create a separate Linear layer for each node type for self-loop
```

Comment:
I think these `loop_weights` uses the feature of each destination node to update its output embeddings. So actually there is no self-loop edges in the graph.

Reply:
Done